Common/include/toolboxes/CIndexedMap.hpp

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <vector>
+#include <unordered_map>
+#include <algorithm>
+
+/*! The CIndexedMap combines a associative map and
+ * and a sequentially indexed container. This allows access to items using a
+ * key (e.g. a string) and at the same time the possibiilty to sequentially loop
+ * over all key/item pairs in a fixed order, based on their insertion.
+ *
+ */
+template <class Key, class Item>
+class CIndexedMap {
+
+public:
+  using KeyType          = Key;
+  using ItemType         = Item;
+  using Map              = std::unordered_map<Key, Item>;
+  using KeyVector        = std::vector<Key>;
+  using InsertionVector  = std::vector<typename Map::iterator>;
+
+protected:
+
+  Map map;
+  InsertionVector insertionVector;
+
+public:
+
+  /*!
+   * \brief Get the size of the container
+   * \return The current number of keys/item pairs.
+   */
+  int GetSize() const { return insertionVector.size(); }
+
+  /*!
+   * \brief Add a new item to the container using a key
+   * \param[in] key - The key associated to the item
+   * \param[in] item - The item that should be stored
+   */
+  void AddItem(const Key& key, const Item& item){
+    insertionVector.emplace_back(map.insert({key, item}).first);
+  }
+
+  /*!
+   * \brief Check if a specific key exists in the container
+   * \param[in] key - The key that should be checked
+   * \return <TRUE> then when the key could be found, otherwise <FALSE>
+   */
+  bool FindKey(const Key& key) const {
+    return (map.count(key) > 0);
+  }
+
+  /*!
+   * \brief Get the key at a specific position
+   * \param[in] i - The position of the key
+   * \return The key at the requested position
+   */
+  const Key& GetKey(const int i) const {
+    return insertionVector[i]->first;
+  }
+
+  /*!
+   * \brief Get the item using the associated key
+   * \param[in] key - The key value
+   * \return The item associated with the specified key
+   */
+  const Item& GetItem(const Key& key) const {
+    return map.at(key);
+  }
+
+  /*!
+   * \brief Get the item at a specific position
+   * \param[in] i - The position of the item
+   * \return The item at the specified position
+   */
+  const Item& GetItem(const int i) const {
+    return insertionVector[i]->second;
+  }
```

Comment:
The code will not compile for `Key = int`, and it will have very surprising behavior if e.g. `Key = unsigned long`, you may want to disambiguate the names e.g. GetItemByIndex, GetNthItem